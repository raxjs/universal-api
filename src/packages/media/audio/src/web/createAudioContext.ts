/* eslint-disable  */
// import { AudioContext } from '../types';
import Events from '@utils/event';

class InnerAudioContext {
  private _audioContext: AudioContext;
  private _source: AudioBufferSourceNode;
  private _gainNode: GainNode;
  private _src = '';
  private _startTime = 0; // time of the audio playback, seconds, user set
  private _playbackTime = 0; // time of the audio playback, seconds
  private _autoplay = false;
  private _buffer: AudioBuffer;
  private _events: any = new Events();
  private _currentTime = 0;
  private _loop = false;
  private _volume = 1;
  private _playbackRate = 1;
  private _eventDeleteCallback = {};
  private _isWaiting = true;
  private _isSeeking = false; // 记录是否正在Seeking
  private _isPlaying = false;
  private _startTimestamp = 0; // timestamp of last playback start, milliseconds
  constructor() {
    // super();

    const AudioContext = window.AudioContext || (window as any).webkitAudioContext;

    this._audioContext = new AudioContext();

    // 创建播放对象节点
    this._source = this._audioContext.createBufferSource();
    // 创建音量节点(如果你需要用调整音量大小的话)
    this._gainNode = this._audioContext.createGain();
  }
  get src() {
    return this._src;
  }
  set src(value) {
    this._src = value;
    this._getData();
  }

  get startTime() {
    return this._startTime;
  }

  set startTime(value) {
    this._startTime = value;
    this._playbackTime = value;
  }

  get autoplay() {
    return this._autoplay;
  }

  set autoplay(value) {
    this._autoplay = value;
  }

  get loop() {
    return this._source?.loop || this._loop;
  }

  set loop(value) {
    this._source.loop = value;
    this._loop = value;
  }

  get duration() {
    return this._buffer?.duration || 0;
  }

  get currentTime() {
    if (this._isPlaying) {
      this._currentTime = (Date.now() - this._startTimestamp)/1000 + this._playbackTime;
    }
    return this._currentTime;
  }

  get paused() {
    return !this._isPlaying;
  }

  get volume() {
    return this._gainNode?.gain?.value || this._volume;
  }

  set volume(value) {
    this._gainNode.gain.value = value;
    this._volume = value;
  }

  get playbackRate() {
    return this._source?.playbackRate?.value || this._playbackRate;
  }

  set playbackRate(value) {
    this._source.playbackRate.value = value;
    this._playbackRate = value;
  }

  _start = (startTime) => {
    // 填充音频buffer数据
    // 创建播放对象节点
    this._source = this._audioContext.createBufferSource();
    // 创建音量节点(如果你需要用调整音量大小的话)
    this._gainNode = this._audioContext.createGain();
    // this.addEventListener('onPaly', this.onPaly);
    this._source.buffer = this._buffer;
    // console.log('buffer', this._buffer);
    this._source.loop = this._loop;
    this._source.playbackRate.value = this._playbackRate;
    this._gainNode.gain.value = this._volume;
    this._source.onended = () => { 
      // console.log('this._isSeeking', this._isSeeking);
      // seek 调用了 source.stop, 会触发onended 事件
      // 此处避开 seek 引发的onended
      if (this._isSeeking) {
        this._isSeeking = false;
        this.currentTime;
        return;
      }
      if (this._isPlaying) this._playbackTime = 0;
      this._isPlaying = false;
      this._events.emit('onEnded');
    }
    // 连接节点对象
    this._source.connect(this._gainNode);
    this._gainNode.connect(this._audioContext.destination);
    this._source.start(0, startTime);
  };

  _getData = () => {
    // console.log('get data start', this);
    this._isWaiting = true;
    this._events.emit('onWaiting');
    const xhr = new XMLHttpRequest();
    xhr.open('GET', this._src, true);
    xhr.responseType = 'arraybuffer';
    xhr.onload = () => {
      this._audioContext.decodeAudioData(xhr.response, (buffer) => {
        this._buffer = buffer;
        this._events.emit('onCanplay');
        this._isWaiting = false;
        this.autoplay && this.play();
      });
    };
    xhr.send();
  };

  play = () => {
    if (this._isPlaying) return;
    this._start(this._playbackTime);
    this._currentTime = this._playbackTime;
    this._startTimestamp = Date.now();
    this._isPlaying = true;
    this._events.emit('onPlay')
  };

  pause = () => {
    this.stop(true);
    this._events.emit('onPause');
  };

  stop = (pause = false) => {
    try {
      if (!this._isPlaying) {
        if (!pause) { // 未播放 stop 时 需要设置为 0
          this._playbackTime = 0;
          this._currentTime = 0;
        }
        return;
      }
      // if (this._isPlaying) {
      this._isPlaying = false;
      this._source.stop(0);
      // }
      this._playbackTime = pause ? (Date.now() - this._startTimestamp)/1000 + this._playbackTime : 0;
      this._currentTime = this._playbackTime;
      this._events.emit('onStop');
      !this._isWaiting && this._events.emit('onCanplay');
    } catch (e) {
      this._events.emit('onError', e);
    }
  };

  seek = async (playbackTime) => {
    try {
      if (playbackTime === undefined) return;
      if (playbackTime > this.duration) {
        this._events.emit('onError', 'Seek time is greater than duration of audio buffer.');
        return;
      }

      this._events.emit('onSeeking');
      if (this._isPlaying) {
        this._isSeeking = true;
        this.stop();
        this._playbackTime = playbackTime;
        this.play();
      } else {
        this._playbackTime = playbackTime;
        this._currentTime = playbackTime;
      }

      this._events.emit('onSeeked');
    } catch (e) {
      this._events.emit('onError', e);
    }
  };

  destroy = () => {
    this._currentTime = this._playbackTime;
    this._isPlaying = false;
    this._audioContext.close()
      .then(() => this._events.emit('onDestroy'))
      .catch((e) => {
        // console.log(e);
        this._events.emit('onError', e);
      });
  };

  onCanplay = (callback = (e) => {}) => {
    this._eventDeleteCallback[callback.toString()] = this._events.register('onCanplay', (e) => {
      // console.log('onCanplay');
      callback(e);
    });
  };

  offCanplay = (callback?: (args?: any) => {}) => {
    if (!callback) {
      this._events.events?.onCanplay?.clear();
      return;
    }
    this._eventDeleteCallback[callback.toString()] && this._eventDeleteCallback[callback.toString()]();
  };

  onPlay = (callback = (e) => {}) => {
    this._eventDeleteCallback[callback.toString()] = this._events.register('onPlay', (e) => {
      // console.log('onPlay');
      callback(e);
    });
  };

  offPlay = (callback?: (args?: any) => {}) => {
    if (!callback) {
      this._events.events?.onPlay?.clear();
      return;
    }
    this._eventDeleteCallback[callback.toString()] && this._eventDeleteCallback[callback.toString()]();
  };

  onPause = (callback = (e) => {}) => {
    this._eventDeleteCallback[callback.toString()] = this._events.register('onPause', (e) => {
      // console.log('onPause');
      callback(e);
    });
  };

  offPause = (callback?: (args?: any) => {}) => {
    if (!callback) {
      this._events.events?.onPause?.clear();
      return;
    }
    this._eventDeleteCallback[callback.toString()] && this._eventDeleteCallback[callback.toString()]();
  };

  onStop = (callback = (e) => {}) => {
    this._eventDeleteCallback[callback.toString()] = this._events.register('onStop', (e) => {
      // console.log('onStop');
      callback(e);
    });
  };

  offStop = (callback?: (args?: any) => {}) => {
    if (!callback) {
      this._events.events?.onStop?.clear();
      return;
    }
    this._eventDeleteCallback[callback.toString()] && this._eventDeleteCallback[callback.toString()]();
  };

  onEnded = (callback = (e) => {}) => {
    this._eventDeleteCallback[callback.toString()] = this._events.register('onEnded', (e) => {
      // console.log('onEnded');
      callback(e);
    });
  };

  offEnded = (callback?: (args?: any) => {}) => {
    if (!callback) {
      this._events.events?.onEnded?.clear();
      return;
    }
    this._eventDeleteCallback[callback.toString()] && this._eventDeleteCallback[callback.toString()]();
  };

  onError = (callback = (e) => {}) => {
    this._eventDeleteCallback[callback.toString()] = this._events.register('onError', (e) => {
      // console.log('onError222');
      callback(e);
    });
  };

  offError = (callback?: (args?: any) => {}) => {
    if (!callback) {
      this._events.events?.onError?.clear();
      return;
    }
    this._eventDeleteCallback[callback.toString()] && this._eventDeleteCallback[callback.toString()]();
  };

  onWaiting = (callback = (e) => {}) => {
    this._eventDeleteCallback[callback.toString()] = this._events.register('onWaiting', (e) => {
      // console.log('onWaiting');
      callback(e);
    });
  };

  offWaiting = (callback?: (args?: any) => {}) => {
    if (!callback) {
      this._events.events?.onWaiting?.clear();
      return;
    }
    this._eventDeleteCallback[callback.toString()] && this._eventDeleteCallback[callback.toString()]();
  };

  onSeeking = (callback = (e) => {}) => {
    this._eventDeleteCallback[callback.toString()] = this._events.register('onSeeking', (e) => {
      // console.log('onSeeking');
      callback(e);
    });
  };

  offSeeking = (callback?: (args?: any) => {}) => {
    if (!callback) {
      this._events.events?.onSeeking?.clear();
      return;
    }
    this._eventDeleteCallback[callback.toString()] && this._eventDeleteCallback[callback.toString()]();
  };

  onSeeked = (callback = (e) => {}) => {
    this._eventDeleteCallback[callback.toString()] = this._events.register('onSeeked', (e) => {
      // console.log('onSeeked');
      callback(e);
    });
  };

  offSeeked = (callback?: (args?: any) => {}) => {
    if (!callback) {
      this._events.events?.onSeeked?.clear();
      return;
    }
    this._eventDeleteCallback[callback.toString()] && this._eventDeleteCallback[callback.toString()]();
  };
}

// 创建audio
const createAudioContext = (): any => {
  let singleInnerAudioContext = new InnerAudioContext();
  return singleInnerAudioContext;
};

export default createAudioContext;
