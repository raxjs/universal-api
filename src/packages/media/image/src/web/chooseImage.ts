import { ChooseImageOptions } from '../types';
import { CONTAINER_NAME } from '@utils/constant';
import { normalize } from '../common';
import chooseMedia from '@uni/video/lib/web/chooseMedia';

const chooseImage = normalize.chooseImage((args: ChooseImageOptions = {}) => {
  const { count = 1, sourceType = ['album', 'camera'], success = () => {}, fail = () => {}, complete = () => {} } = args;
  chooseMedia({
    count: 1,
    accept: 'image/*',
  }).then((res) => {
    const result = {
      tempFilePaths: res.tempFiles.map((item) => item.tempFilePath),
      tempFiles: res.tempFiles.map((item) => ({
        path: item.tempFilePath,
        size: item.size,
      })),
      files: res.files,
    };
    success(result as any);
    complete(result as any);
  }).catch((e) => {
    fail(e);
    complete(e);
  });
}, CONTAINER_NAME.WEB);

export default chooseImage;
