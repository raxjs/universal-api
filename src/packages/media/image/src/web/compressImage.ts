const compressImage = () => {
  throw new Error('Uni API: compressImage不支持');
};

export default compressImage;
