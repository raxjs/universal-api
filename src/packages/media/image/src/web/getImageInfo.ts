import { normalize } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

const getImageInfo = normalize.getImageInfo((args) => {
  const image = new Image();
  image.src = args.src;
  if (image.naturalWidth) {
    const res = {
      width: image.naturalWidth,
      height: image.naturalHeight,
      path: args.src,
    };
    args.success(res);
    args.complete(res);
  } else {
    const setImageInfo = () => {
      const res = {
        width: image.width,
        height: image.height,
        path: args.src,
      };
      args.success(res);
      args.complete(res);
    };
    if (image.complete) {
      setImageInfo();
    } else {
      let checkTimer = null;
      const clear = () => {
        clearInterval(checkTimer);
        checkTimer = null;
      };
      checkTimer = setInterval(() => {
        if (image.width > 0 || image.height > 0) {
          setImageInfo();
          clear();
        }
      }, 40);
      image.onload = () => {
        if (checkTimer) {
          setImageInfo();
          clear();
        }
      };
      // eslint-disable-next-line no-multi-assign
      image.onabort = image.onerror = (e) => {
        args.fail(e);
        args.complete(e);
        clear();
      };
    }
  }
}, CONTAINER_NAME.WEB);

export default getImageInfo;
