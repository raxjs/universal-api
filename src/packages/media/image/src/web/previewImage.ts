import { normalize } from '../common';
import { CONTAINER_NAME } from '@utils/constant';
import { PreviewImageOptions } from '../types';

const clsPrefix = '__universal_preview_image';

const styles = `.${clsPrefix} {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 9999;
  text-align: center;
  background: #000;
  box-sizing: border-box;
  outline: none;
  overflow: hidden;
  animation-duration: .3s;
  animation-fill-mode: both;
}
.${clsPrefix}_pageindex {
  position: absolute;
  top: 12px;
  left: 0;
  right: 0;
  text-align: center;
  color: #fff;
  font-size: 16px;
  z-index: 2;
}
.${clsPrefix}_swiper {
  position: relative;
  width: 100%;
  height: 100%;
  z-index: 1;
  display: flex;
  align-items: center;
  transition-property: transform;
  box-sizing: content-box;
}
.${clsPrefix}_item {
  max-height: 70%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-shrink: 0;
  width: 100%;
  position: relative;
  transition-property: transform;
  overflow: hidden;
}
.${clsPrefix}_img {
  width: 100%;
}
@keyframes ${clsPrefix}_fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
@keyframes ${clsPrefix}_fadeOut {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}`.replace(/\n/g, '');

let styleElement: HTMLElement | null = null;

const previewImage = normalize.previewImage((args: PreviewImageOptions) => {
  const { urls = [] } = args || {};
  let { current = 0 } = args || {};

  if (Object.prototype.toString.call(urls) !== '[object Array]') {
    throw new Error('the arguments must contain urls that type is a Array');
  }

  if (urls.length <= 0) {
    throw new Error('the urls`s length is at least 1');
  }

  if (!Number.isSafeInteger(current)) {
    throw new Error('the current must be a safe integer');
  }

  current = Math.min(Math.max(current, 0), urls.length - 1);

  if (!styleElement) {
    // create a style tag for keyframes
    styleElement = document.createElement('style');
    styleElement.innerHTML = styles;
    document.body.appendChild(styleElement);
  }

  let containerEle = document.createElement('div');
  containerEle.className = clsPrefix;
  containerEle.style.animationName = `${clsPrefix}_fadeIn`;

  containerEle.addEventListener('click', () => {
    containerEle.style.animationName = `${clsPrefix}_fadeOut`;
    setTimeout(() => {
      document.body.removeChild(containerEle);
      containerEle = null;
    }, 300);
  });

  const pageIndexEle = document.createElement('div');
  pageIndexEle.className = `${clsPrefix}_pageindex`;
  pageIndexEle.innerText = `${current + 1}/${urls.length}`;
  containerEle.appendChild(pageIndexEle);

  const swiperEle = document.createElement('div');
  swiperEle.className = `${clsPrefix}_swiper`;

  const damp = 1;
  const { clientWidth } = document.documentElement;
  let startX = 0;
  swiperEle.addEventListener('touchstart', (e) => {
    if (e.touches.length > 1) {
      e.stopPropagation();
      e.preventDefault();
    }
    startX = e.targetTouches[0].pageX;
  });
  swiperEle.addEventListener('touchmove', (e) => {
    if (e.touches.length > 1) {
      e.stopPropagation();
      e.preventDefault();
    }
    const moveX = e.targetTouches[0].pageX - startX;
    swiperEle.style.transform = `translateX(${-clientWidth * current + moveX * damp}px)`;
  });
  swiperEle.addEventListener('touchend', (e) => {
    const endX = e.changedTouches[0].pageX - startX;
    if (Math.abs(endX * damp) > clientWidth / 2) {
      current = endX < 0 ? Math.min(current + 1, urls.length - 1) : Math.max(current - 1, 0);
    }
    pageIndexEle.innerText = `${current + 1}/${urls.length}`;
    const transitionDuration = 300;
    swiperEle.style.transitionDuration = `${transitionDuration}ms`;
    swiperEle.style.transform = `translateX(${-clientWidth * current}px)`;
    setTimeout(() => {
      swiperEle.style.transitionDuration = '0ms';
    }, transitionDuration);
  });
  swiperEle.style.transform = `translateX(${-clientWidth * current}px)`;
  containerEle.appendChild(swiperEle);

  urls.forEach((url) => {
    const swiperItemEle: any = document.createElement('div');
    swiperItemEle.className = `${clsPrefix}_item`;
    swiperItemEle.addEventListener('click', (e) => {
      e.stopPropagation();
    });
    swiperEle.appendChild(swiperItemEle);

    const imageEle = document.createElement('img');
    imageEle.className = `${clsPrefix}_img`;
    imageEle.src = url;
    swiperItemEle.appendChild(imageEle);

    swiperItemEle.addEventListener('touchstart', (e) => {
      if (e.touches.length > 1) {
        e.stopPropagation();
        e.preventDefault();
        const point1 = e.touches[0];
        const point2 = e.touches[1];
        const xLen = Math.abs(point2.pageX - point1.pageX);
        const yLen = Math.abs(point2.pageY - point1.pageY);
        swiperItemEle.touchDistance = Math.sqrt(xLen * xLen + yLen * yLen);
      } else {
        swiperItemEle.touched = {
          x: e.touches[0].pageX,
          y: e.touches[0].pageY,
        };
      }
    });
    swiperItemEle.addEventListener('touchmove', (e) => {
      if (e.touches.length > 1) {
        e.stopPropagation();
        e.preventDefault();
        const xLen = Math.abs(e.touches[0].pageX - e.touches[1].pageX);
        const yLen = Math.abs(e.touches[1].pageY - e.touches[1].pageY);
        const touchDistance = Math.sqrt(xLen * xLen + yLen * yLen);

        if (swiperItemEle.touchDistance) {
          const pinchScale = (swiperItemEle.previousPinchScale || 1) + (touchDistance - swiperItemEle.touchDistance) / swiperItemEle.touchDistance;
          const imageScale = Math.max(1, pinchScale);
          imageEle.style.transform = `scale(${imageScale})`;
          swiperItemEle.previousPinchScale = imageScale;
        }
        swiperItemEle.touchDistance = touchDistance;
      } else if (swiperItemEle.previousPinchScale > 1) {
        e.stopPropagation();
        const x = (swiperItemEle.preX || 0) + (e.touches[0].pageX - swiperItemEle.touched.x);
        const y = (swiperItemEle.preY || 0) + (e.touches[0].pageY - swiperItemEle.touched.y);
        imageEle.style.transform = `scale(${swiperItemEle.previousPinchScale}) translate(${x}px, ${y}px)`;
        swiperItemEle.preX = x;
        swiperItemEle.preY = y;
        swiperItemEle.touched = {
          x: e.touches[0].pageX,
          y: e.touches[0].pageY,
        };
      }
    });
  });

  document.body.appendChild(containerEle);
}, CONTAINER_NAME.WEB);

export default previewImage;
