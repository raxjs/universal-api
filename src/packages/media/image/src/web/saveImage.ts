const saveImage = () => {
  throw new Error('Uni API: saveImage不支持');
};

export default saveImage;
