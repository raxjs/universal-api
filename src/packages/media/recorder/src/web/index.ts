/* eslint-disable  */

const getUserMedia = (constraints) => {
  const navigate: any = navigator;
  const rawGetUserMedia = (navigator.mediaDevices && navigator.mediaDevices.getUserMedia) || navigate.getUserMedia || navigate.webkitGetUserMedia || navigate.mozGetUserMedia;

  if (!rawGetUserMedia) {
    return Promise.reject(new Error('getUserMedia is not implemented in this browser'));
  }

  return rawGetUserMedia.call((navigator.mediaDevices || navigator), constraints);
};

const getAudioMedia = () => {
  return getUserMedia({ audio: true });
}

function recorderManger () {
  this.duration = 60000;
  this.sampleRate = 8000;
  this.numberOfChannels = 2;
  this.encodeBitRate = 48000;
  this.format = 'aac';
  this.frameSize = -1;
  this.mediaRecorder = null;
  this.chunks = [];
  this.curDuration = 0;
  this.timer = null;
  this.events = {};
}

recorderManger.prototype.start = function (args) {
  if (args && typeof args === 'object') {
    for (const key in args) {
      if (Object.prototype.hasOwnProperty.call(args, key)) {
        this[key] = args[key];
      }
    }
  }
  getAudioMedia().then(stream => {
    this.mediaRecorder = new MediaRecorder(stream);

    this.mediaRecorder.onstart = this.events.onstart;
    this.mediaRecorder.onpause = this.events.onpause;
    this.mediaRecorder.onresume = this.events.onresume;
    this.mediaRecorder.onstop = this.events.onstop;
    this.mediaRecorder.ondataavailable = this.events.ondataavailable;
    this.mediaRecorder.onerror = this.events.onerror;
    
    if (this.frameSize > 0) {
      this.mediaRecorder.start(this.frameSize);
    } else {
      this.mediaRecorder.start();
    }
  }).catch(() => {
    throw new Error('@uni/apis: getRecorderManager不支持');
  });
}

recorderManger.prototype.pause = function () {
  this.mediaRecorder && this.mediaRecorder.state === 'recording' && this.mediaRecorder.pause();
}

recorderManger.prototype.resume = function () {
  this.mediaRecorder && this.mediaRecorder.state === 'paused' && this.mediaRecorder.resume();
}

recorderManger.prototype.stop = function () {
  this.mediaRecorder && this.mediaRecorder.state === 'recording' && this.mediaRecorder.stop();
}

recorderManger.prototype.onStart = function (callback = () => {}) {
  this.events.onstart = () => {
    this.clear();

    this.timer = setInterval(() => {
      if (this.mediaRecorder && this.mediaRecorder.state === 'recording') {
        this.curDuration += 500;

        if (this.curDuration >= this.duration) {
          this.curDuration -= 500;
          this.stop();
        }
      }
    }, 500);

    callback();
  };
}

recorderManger.prototype.onPause = function (callback = () => {}) {
  this.events.onpause = callback;
}

recorderManger.prototype.onResume = function (callback = () => {}) {
  this.events.onresume = callback;
}

recorderManger.prototype.onStop = function (callback = (res) => {}) {
  this.events.onstop = () => {
    const blob = new Blob(this.chunks, { 'type' : 'audio/ogg; codecs=opus' });
    const tempFilePath = window.URL.createObjectURL(blob);
    const fileSize = blob.size;
    const duration = this.curDuration + 500;
    this.clear();
    callback({
      tempFilePath,
      fileSize,
      duration
    });
  };
}

recorderManger.prototype.onFrameRecorded = function (callback = (data) => {}) {
  this.events.ondataavailable = (e) => {
    this.chunks.push(e.data);
    callback(e.data);
  };
}

recorderManger.prototype.onError = function (callback = (e) => {}) {
  this.events.onerror = (e) => {
    this.clear();
    callback(e);
  };
}

recorderManger.prototype.clear = function () {
  this.chunks = [];
  clearInterval(this.timer);
  this.timer = null;
  this.curDuration = 0;
};

const singleRecorderManager = new recorderManger();

// 使用单例返回全局唯一的录音管理器
const getRecorderManager = () => {
  return singleRecorderManager;
} 

export default getRecorderManager;
