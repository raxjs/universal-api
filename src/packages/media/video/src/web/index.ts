import chooseVideo from './chooseVideo';
import createVideoContext from './createVideoContext';
import chooseMedia from './chooseMedia';

export {
  chooseVideo,
  createVideoContext,
  chooseMedia,
};

export default {
  chooseVideo,
  createVideoContext,
  chooseMedia,
};
