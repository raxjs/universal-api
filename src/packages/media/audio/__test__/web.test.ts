import { sleep, testWebAPI } from '@/utils/__test__/util';
import { DOMUtil } from '@/utils/__test__/web';

testWebAPI('crateAudioContext', async () => {
  window.AudioContext = jest.fn().mockImplementation(() => {
    return {
      createBufferSource: () => {},
      createGain: () => {},
    };
  });

  const { createAudioContext } = require('../src/index');
  const res = createAudioContext();
  await sleep(300);
  expect(document.body).toMatchSnapshot();
});
