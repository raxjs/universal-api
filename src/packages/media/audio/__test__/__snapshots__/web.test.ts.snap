// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Platform API: crateAudioContext Test container: web 1`] = `<body />`;
